### PR TITLE
Fix incorrect recovery for unicode general category values in regex

### LINKED ISSUE #39985: [Bug]: RegExp parser recovers incorrectly when there are incorrect abbreviation letters in a Unicode general category value
### Description

Let's take the pattern `\p{Lz}` as an example. In here parser should report the incorrect token `z` and continue parsing. This way there won't be any unmeaning full diagnostics as shown below.

Screenshot 2023-03-24 at 11 27 25

In order to solve this we need to report the diagnostic in `processAbbrWithLetter` method. This should applied to other Unicode general categories as well.

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### LINKED ISSUE #40490: [Bug]: Invalid recovery for Regex general unicode category
### Description

```
public function main() {
    string:RegExp a = re `\p{Lz}`;
}
```

for the above code Ballerina parser recover as 

```
public function main() {
    string:RegExp a = re `\p{L MISSING[}] MISSING[\ INVALID[z]] MISSING[\ INVALID[}]]`;
}
```

But after the `MISSING[}]`, `MISSING[\ INVALID[z]]` should not be occurred. Because z is valid after the `\p{L}`.

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by SasinduDilshara**
Fixes #39985 
Fixes #40490

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/RegExpLexer.java [hunk 1]
```diff
@@ -364,9 +364,20 @@ private void processAbbrWithOther() {
...
                 this.reader.advance();
-                break;
+                return;
             default:
-                break;
+                checkInvalidTokenInUnicodeGeneralCategory();
+        }
+    }
+
+    private void checkInvalidTokenInUnicodeGeneralCategory() {
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/RegExpLexer.java:373]: I think we dont need this change. we can simply move the default block in the switch of the `processReUnicodeGeneralCategoryAbbr()` method out and place it after the switch to get the things done

**Comment by SasinduDilshara** [comment]
The approach we try to resolve #39985 is got changed.
Planning to fix #40490 and fix this #39985 with same approach.

**Comment by SasinduDilshara** [comment]
> The approach we try to resolve #39985 is got changed. Planning to fix #40490 and fix this #39985 with same approach.
> 
> Hence closing the PR

had a discussion with @rdulmina and came to a conclusion to handle the original issue using this PR.

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/RegExpLexer.java (hunk 1 above)
Comment (SasinduDilshara) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/RegExpLexer.java:373]: Will check

Comment (SasinduDilshara) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/RegExpLexer.java]: Updated, Please dcheck

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/RegExpLexer.java [hunk 2]
```diff
@@ -269,12 +269,15 @@ private STToken processReUnicodeGeneralCategoryAbbr() {
...
-                reportLexerError(DiagnosticErrorCode.ERROR_INVALID_TOKEN_IN_REG_EXP);
+                break;
         }
 
+        if (!isEndOfUnicodePropertyEscape()) {
+            while (!isEndOfUnicodePropertyEscape()) {
+                this.reader.advance();
+            }
+            reportLexerError(DiagnosticErrorCode.ERROR_INVALID_TOKEN_IN_REG_EXP);
+        }
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/RegExpLexer.java:280]:
```suggestion
        }
        
```

**Comment by rdulmina** [comment]
Let's send this to `7.x` branch as well

**Review by rdulmina — APPROVED** [review]

**Pull request merged**